gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/MultiLongWatermark.java

```diff
@@ -0,0 +1,159 @@
+/* (c) 2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.source.extractor.extract.kafka;
+
+import java.math.RoundingMode;
+import java.util.List;
+
+import com.beust.jcommander.internal.Lists;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ComparisonChain;
+import com.google.common.math.LongMath;
+import com.google.common.primitives.Shorts;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+
+import gobblin.source.extractor.Watermark;
+
+
+/**
+ * A {@link gobblin.source.extractor.Watermark} that holds multiple long values.
+ *
+ * @author ziliu
+ */
+public class MultiLongWatermark implements Watermark {
+
+  private static final Gson GSON = new Gson();
+
+  private final List<Long> values;
+
+  /**
+   * Copy constructor.
+   */
+  public MultiLongWatermark(MultiLongWatermark other) {
+    this.values = Lists.newArrayList(other.values);
+  }
+
+  public MultiLongWatermark() {
+    this.values = Lists.newArrayList();
+  }
+
+  /**
+   * Increment the idx'th value. idx must be between 0 and size()-1.
+   */
+  public void increment(int idx) {
+    Preconditions.checkElementIndex(idx, this.values.size());
+    Preconditions.checkArgument(this.values.get(idx) < Long.MAX_VALUE);
+    this.values.set(idx, this.values.get(idx) + 1);
+  }
+
+  /**
+   * A MultiLongWatermark w1 is less than w2, if there exists i, such that
+   * the 0th to (i-1)th values of w1 are equivalent as those of w2, and the ith
+   * value of w1 is less than the that of w2.
+   *
+   * @param wm must be an instance of MultiLongWatermark, and this.size() must
+   * equal wm.size().
+   */
+  @Override
+  public int compareTo(Watermark wm) {
```

Comment:
The fact that `Watermark` currently uses `Comparable` isn't leveraged anywhere. In fact the `calculatePercentCompletion` method isn't invoked anywhere either. It was deemed too much additional work to modify the framework as this point in time in order to integrate with these methods. However, I do think that keeping the `Watermark` interface `Comparable` has advantages. I think that `Watermark`s should actually be sortable. In most data ingestion use cases, one typically has a low watermark and a high watermark, and a single execution pulls data from the low watermark to the high watermark. The next execution takes the high watermark of the previous run, and uses it to calculate a new low and high watermark. For example, a MySQL puller may pull data based on the value of a SystemModifiedTimestamp column. Its first execution may pull data from [0, x = current epoch time), the next execution will pull from [x, current epoch time), etc; where the watermark is a timestamp.

I think the actual `compareTo` method may come in handy in the `WatermarkInterval` class, where a user constructs a `WatermarkInterval` that spans from a low watermark to a high watermark. Each `WorkUnit` has a single `WatermarkInterval` assigned to it. The framework can compare the low watermark and the high watermark using the `compareTo` method in order to check that one is actually lower than the other.

Here is another use case: the `Extractor` wants to pull data from a low watermark to a high watermark, but doesn't want to worry about tracking when its current watermark is >= the high watermark. By using the `compareTo` method the framework can handle this logic for the `Extractor`. This way the `Extractor` is just concerned about pulling data, but doesn't need to worry about tracking when it needs to stop pulling data.